pkg/solana/txm/txm.go

```diff
@@ -718,41 +735,58 @@ func (txm *Txm) simulateTx(ctx context.Context, tx *solanaGo.Transaction) (res *
+	if resErr != nil {
 		// handle various errors
 		// https://github.com/solana-labs/solana/blob/master/sdk/src/transaction/error.rs
-		errStr := fmt.Sprintf("%v", res.Err) // convert to string to handle various interfaces
+		errStr := fmt.Sprintf("%v", resErr) // convert to string to handle various interfaces
+		txm.lggr.Info(errStr)
 		logValues := []interface{}{
-			"id", id,
 			"signature", sig,
-			"result", res,
+			"error", resErr,
+		}
+		// return TxFailRevert on any error if when processing error during confirmation
+		errType := TxFailRevert
+		// return TxFailSimRevert on any known error when processing simulation error
+		if simulation {
+			errType = TxFailSimRevert
 		}
 		switch {
 		// blockhash not found when simulating, occurs when network bank has not seen the given blockhash or tx is too old
 		// let confirmation process clean up
 		case strings.Contains(errStr, "BlockhashNotFound"):
-			txm.lggr.Debugw("simulate: BlockhashNotFound", logValues...)
-		// transaction will encounter execution error/revert, mark as reverted to remove from confirmation + retry
-		case strings.Contains(errStr, "InstructionError"):
-			_, err := txm.txs.OnError(sig, txm.cfg.TxRetentionTimeout(), TxFailSimRevert) // cancel retry
-			if err != nil {
-				logValues = append(logValues, "stateTransitionErr", err)
+			txm.lggr.Debugw("BlockhashNotFound", logValues...)
+			// return no failure for this error when simulating to allow later send/retry code to assign a proper blockhash
+			// in case the one provided by the caller is outdated
+			if simulation {
+				return txState, NoFailure
 			}
-			txm.lggr.Debugw("simulate: InstructionError", logValues...)
-		// transaction is already processed in the chain, letting txm confirmation handle
+			return Errored, errType
+		// transaction will encounter execution error/revert
+		case strings.Contains(errStr, "InstructionError"):
+			txm.lggr.Debugw("InstructionError", logValues...)
+			return Errored, errType
+		// transaction is already processed in the chain
 		case strings.Contains(errStr, "AlreadyProcessed"):
-			txm.lggr.Debugw("simulate: AlreadyProcessed", logValues...)
+			txm.lggr.Debugw("AlreadyProcessed", logValues...)
+			// return no failure for this error when simulating in case there is a race between broadcast and simulation
+			// when doing both in parallel
+			if simulation {
+				return txState, NoFailure
+			}
+			return Errored, errType
 		// unrecognized errors (indicates more concerning failures)
 		default:
-			_, err := txm.txs.OnError(sig, txm.cfg.TxRetentionTimeout(), TxFailSimOther) // cancel retry
-			if err != nil {
-				logValues = append(logValues, "stateTransitionErr", err)
+			// if simulating, return TxFailSimOther if error unknown
+			if simulation {
+				errType = TxFailSimOther
 			}
-			txm.lggr.Errorw("simulate: unrecognized error", logValues...)
+			txm.lggr.Errorw("unrecognized error", logValues...)
+			return Errored, errType
```

Comment:
should this be fatally errored if it's simulation type?

Reply:
Good point, I'll check with CCIP to see what they think. I was sort of following the EVM TXM philosophy of keeping unknown errors retryable but surfacing it through logs. But we can tackle this with the later update for fatal errors